exonum/src/storage/indexes_metadata.rs

```diff
@@ -154,8 +161,41 @@ pub fn assert_index_type(name: &str, index_type: IndexType, is_family: bool, vie
     }
 }
 
+#[derive(Debug, Clone)]
+pub(crate) enum VersionStatus {
+    Unsupported {
+        core_ver: Version,
+        storage_ver: Version,
+    },
+    Unspecified {
+        core_ver: Version,
+    },
+    Supported(Version),
+}
+
+pub(crate) fn storage_version_status<T: AsRef<Snapshot>>(view: T) -> VersionStatus {
+    let metadata = BaseIndex::indexes_metadata(view);
+    match metadata.get::<_, Version>(CORE_STORAGE_VERSION_KEY) {
+        Some(ver) if ver == CORE_STORAGE_VERSION => VersionStatus::Supported(ver),
+        Some(ver) => VersionStatus::Unsupported {
+            core_ver: CORE_STORAGE_VERSION,
+            storage_ver: ver,
+        },
+        None => VersionStatus::Unspecified {
+            core_ver: CORE_STORAGE_VERSION,
+        },
+    }
+}
+
+// Sets storage version in metadata index and returns the version set.
+pub(crate) fn initialize_storage_version(view: &mut Fork) -> Version {
```

Comment:
The idea behind `VersionStatus` is to express the result of comparison of the value in the storage and the value in the core. I think it makes sense that we don't `write` `VersionStatus`, but rather we write the `Version`.

Reply:
It *could* make sense to combine functionality of both into one class, e.g., to provide stronger guarantees about the range of supported versions than `u64`:

```rust
pub enum Version {
    /// Initial version of the storage used in Exonum >=0.8.
    Version1,
    /*
    /// Hypothetical new version. Changes compared to version 1 may be
    /// documented here.
    Version2,
    */
    Unsupported(u64),
}
```

Reply:
I can definitely see that, I like the idea of making `Version` a dedicated type instead of `u64`. The only issue I have with this approach are the variant names (`Version1`, `Version2`, `Version3` seems hacky). Also, since this version is internal for the core, do we really need to document all these versions? For now I will settle with
```rust
pub(crate) enum Version {
    Supported(u64),
    Unsupported(u64),
    Unspecified,
}
```

Reply:
More intelligent naming scheme may make sense, although I cannot come up with it right away. Specifying versions explicitly has the added benefit compared to `Supported(u64)` in that *all* supported versions are explicitly enumerated (there should be relatively few of them, I hope), so there is no chance to create something like `Version::Supported(100500)`. As for comments, I think they would be useful at least for core developers.